### PR TITLE
OPENTOK-51006 - 	 Create a Captions API sample app

### PR DESCRIPTION
**Pull request by behei-vonage**
This PR adds a sample app for captions API for the client side.

To test this app:
* Make sure you have a server side running - this can be found at https://github.com/opentok/learning-opentok-node.
* In Basic-Captions/js/config.js, change the URL to `http://localhost:8080`
* In Basic-Captions folder, start a local development server, such as `python3 -m http.server` or whatever you prefer.
* Navigate to `localhost:8000` from two tabs
* Open console logs
* Start the captions by clicking a button and start talking - notice the captions flowing.
* Stop captions from either one of the tabs.
* Notice no errors related to captions (you might see an error but that is related to missing `favicon.ico`) 

Note: the server side sample will be added separately to this repo - https://github.com/opentok/learning-opentok-node.

### CODE REVIEW

File: Basic-Captions/README.md
Comment (michaelwhittemore) [inline Basic-Captions/README.md:7]: typo 'an' to 'and'

File: Basic-Captions/js/app.js
Comment (michaelwhittemore) [inline Basic-Captions/js/app.js:3]: super nit picky, buy given that this is customer facing shouldn't we use `const` & `let`?

**Comment by michaelwhittemore** [comment]
Has this been tested? It seemed like there was an issue earlier when publishing and subscribing in the same tab?

File: Basic-Captions/js/app.js
Comment (behei-vonage) [reply Basic-Captions/js/app.js:3]: good point, I just copy-pasted the skeleton from a different sample app. will update 🚀

**Comment by behei-vonage** [comment]
> Has this been tested? It seemed like there was an issue earlier when publishing and subscribing in the same tab?

@michaelwhittemore, haven't done testing yet, still need a server component 🙏

File: Basic-Captions/js/app.js
Comment (michaelwhittemore) [inline Basic-Captions/js/app.js:29]:
```suggestion
    const subscriberOptions = {
```

File: Basic-Captions/js/app.js
Comment (michaelwhittemore) [inline Basic-Captions/js/app.js:25]:
```suggestion
  const session = OT.initSession(apiKey, sessionId);
```

File: Basic-Captions/js/app.js
Comment (michaelwhittemore) [inline Basic-Captions/js/app.js:57]:
```suggestion
      const publisherOptions = {
```

**Comment by michaelwhittemore** [comment]
> > Has this been tested? It seemed like there was an issue earlier when publishing and subscribing in the same tab?
> 
> @michaelwhittemore, haven't done testing yet, still need a server component 🙏

Makes sense, in that case I'd like to hold off approving until we have a server component to interop with if that's ok. If that's a blocker I can approve now.

**Comment by behei-vonage** [comment]
> > > Has this been tested? It seemed like there was an issue earlier when publishing and subscribing in the same tab?
> > 
> > 
> > @michaelwhittemore, haven't done testing yet, still need a server component 🙏
> 
> Makes sense, in that case I'd like to hold off approving until we have a server component to interop with if that's ok. If that's a blocker I can approve now.

that makes sense, let's wait until there is a server component 🙏

**Comment by devwithzachary** [comment]
I've also requested @conshus review this sample as this repository is managed by the DevRel team, please let him know when it is ready to rest and review. Thank you!

**Comment by conshus** [comment]
Hi all. 
When this is ready, is it cool if I go over it and do some refactoring?
Noticed a couple of things at a first glance:
- In the readme it shows one way of getting the credentials (which is inline with the other web examples), but then app.js does something different.
- This demo relies on the person having a server running instead of just passing in the Credentials
- Could probably remove the jQuery dependency with some small changes
- The index.html links to the Archiving sample

Like I said, if it's cool, once the functionality it complete, I can go in and refactor some things to make it more similar to the other Web Samples.

Thanks.

**Comment by behei-vonage** [comment]
> Hi all. When this is ready, is it cool if I go over it and do some refactoring? Noticed a couple of things at a first glance:
> 
> * In the readme it shows one way of getting the credentials (which is inline with the other web examples), but then app.js does something different.
> * This demo relies on the person having a server running instead of just passing in the Credentials
> * Could probably remove the jQuery dependency with some small changes
> * The index.html links to the Archiving sample
> 
> Like I said, if it's cool, once the functionality it complete, I can go in and refactor some things to make it more similar to the other Web Samples.
> 
> Thanks.

Hi @conshus,

Thanks so much for all the suggestions 🙏 I believe that this is ready for review as this sample works with the learning-opentok-node sample app changes. 

If you don't mind helping out with the suggestions you made, I'd greatly appreciate it. 

Thanks!

**Comment by behei-vonage** [comment]
@conshus, I believe I addressed most of your feedback on this one. Though, I ran into issues when migrating from ajax to fetch. I was able to get session and token generating to work; however, I keep on getting an empty body when I try to make a request to `/archive/start.` So I kept ajax as is for now, maybe you'd like to give it a try. Thanks!

**Comment by conshus** [comment]
@behei-vonage I'll take a look. For the Captions API sample, you are also archiving? I don't see that in the code. Where are you calling `/archive/start`?

**Comment by behei-vonage** [comment]
> @behei-vonage I'll take a look. For the Captions API sample, you are also archiving? I don't see that in the code. Where are you calling `/archive/start`?

This sample doesn't call any of the Archiving methods. I updated the index.html to not link to Archiving branch and removed a reference to it in the readme.md file

**Comment by behei-vonage** [comment]
We might consider some limitations with this sample app: 

* publisher A joins the session, the captions are started via an Ajax request to the backend, and the client-side captionId is assigned the value that came from the server-side
* publisher B joins the session; it tries to request to start the captions; however, it gets the error 500 with the “captions already enabled” message. The client-side captionId doesn’t get the value from the server
* now, if I try to stop the captions from publisher A, I have no issues. 
* However, if I try to stop them from publisher B, I cannot since the client-side captionsId wasn’t initialized and therefore, the Ajax request from the client to the backend fails.

One way this limitation can be overcome is with the client-side API requesting whether the captions are enabled and grab the captions ID. From discussion on [this thread](https://vonage.slack.com/archives/C0287LFFD0F/p1678143525080249), this might make it in for development in Q2. 

Another way I can think of has a database that stores the captionIds but I feel like it is getting too complex at this point and out of the scope of this Basic-Captions sample app.

cc: @v-kpheng so that you are in the loop and maybe have some suggestions on how to proceed

**Comment by v-kpheng** [comment]
> We might need to consider some limitations with this sample app:
> 
> * publisher A joins the session, the captions are started via an Ajax request to the backend, and the client-side captionId is assigned the value that came from the server-side
> * publisher B joins the session; it tries to request to start the captions; however, it gets the error 500 with the “captions already enabled” message. The client-side captionId doesn’t get the value from the server
> * now, if I try to stop the captions from publisher A, I have no issues.
> * However, if I try to stop them from publisher B, I cannot since the client-side captionsId wasn’t initialized and therefore, the Ajax request from the client to the backend fails.
> 
> One way this limitation can be overcome is with the client-side API requesting whether the captions are enabled and grab the captions ID. From discussion on [this thread](https://vonage.slack.com/archives/C0287LFFD0F/p1678143525080249), this might make it in for development in Q2.
> 
> Another way I can think of has a database that stores the captionIds but I feel like it is getting too complex at this point and out of the scope of this Basic-Captions sample app.
> 
> cc: @v-kpheng so that you are in the loop and maybe have some suggestions on how to proceed

Thanks, @behei-vonage.  We can always enhance in Q2, once those APIs are completed.  The sample app you created will help unblock our users.

**Comment by behei-vonage** [comment]
awesome @michaelwhittemore helped me out with finding a solution that would work in a 1-to-1 session 💪 

that required some changes to the server side that were done in the following pull request - https://github.com/opentok/learning-opentok-node/pull/40. It'd be great if someone from DevRel could take a look at it. 

@conshus, while it would be great for this sample to look as close to the Archiving samples, there are currently some limitations to what methods we have available. For example, there is a session-level event `archiveStarted`; however, there is not a session-level even `captionsStarted` This method could have been used for retrieving the captionsId before sending a stop request but that is not currently possible; therefore, the stop route on the server side had to be modified to `/captions/stop` instead of /captions/:captionsId/stop`

Please review this sample app when you have a chance and let me know your thoughts. Thanks!

File: Basic-Captions/js/app.js
Comment (michaelwhittemore) [inline Basic-Captions/js/app.js:84]: Super minor nit pick, but we may want to consider breaking this conditional into multiple lines so that it looks cleaner i.e. 
```
if (oldCaptionBox) {
    oldCaptionBox.remove();
}

File: Basic-Captions/README.md
Comment (michaelwhittemore) [inline Basic-Captions/README.md:43]: I think we should mention that starting captions requires a moderator token

File: Basic-Captions/css/app.css
Comment (michaelwhittemore) [inline Basic-Captions/css/app.css:39]: I think the caption looked a little small when we were troubleshooting, but I was only seeing it via screenshare so I will leave the judgement call up to you. It's obviously not important or a blocker at all.

Comment (behei-vonage) [reply Basic-Captions/css/app.css:39]: I moved the publisher to the top left instead of the bottom left so that the Publisher object does not block the captions as they come in, but I kept the font size the same; I feel like it's big enough

Comment (michaelwhittemore) [reply Basic-Captions/css/app.css:39]: Awesome, thanks man

**Review by michaelwhittemore — APPROVED** [review]
LGTM, great job

**Comment by behei-vonage** [comment]
@conshus could I please have your help in reviewing this one? thanks!

File: Basic-Captions/js/app.js
Comment (michaelwhittemore) [inline Basic-Captions/js/app.js:34]: Should we include `captions started with ID:`?

File: Basic-Captions/js/app.js
Comment (behei-vonage) [reply Basic-Captions/js/app.js:106]: I don't think it does actually. But this was a copy-pasta from other sample apps so I'll just keep it as-is

**Comment by behei-vonage** [comment]
Note: there were a lot of awesome updates by @conshus that were carried over to this PR from his branch - https://github.com/opentok/opentok-web-samples/tree/opentok-51006-refactor/Basic-Captions. I couldn't think of a faster way to bring those changes over here except for copying it over, but the credit for that goes over to Dwane. Thanks for helping out with this 💪

**Comment by conshus** [comment]
No problem @behei-vonage . Glad I could help.  Great job! 👍🏾

**Comment by dwivedisachin** [comment]
Tested LGTM!! 🚀🚀

**Comment by dwivedisachin** [comment]
Tested LGTM!! 🚀🚀

**Pull request merged**
